lib/Sema/CSBindings.cpp

```diff
@@ -173,8 +173,10 @@ void ConstraintSystem::PotentialBindings::addPotentialBinding(
   // check whether we can combine it with another
   // supertype binding by computing the 'join' of the types.
   if (binding.Kind == AllowedBindingKind::Supertypes &&
-      !binding.BindingType->hasTypeVariable() && !binding.DefaultedProtocol &&
-      !binding.isDefaultableBinding() && allowJoinMeet) {
+      !binding.BindingType->hasTypeVariable() &&
+      !binding.BindingType->hasUnboundGenericType() &&
```

Comment:
It looks like the diff here is checking for `hasUnboundGenericType()`.

It's not clear to me why we would have to do that here. It's possible `join` was hitting an `llvm_unreachable`, but I recently fixed that to return `Unimplemented` in the cases we aren't currently handling - it was an oversight that I checked in the `llvm_unreachable`.

Reply:
I've added that because `Kickstarter-Prelude` has hitting "non-canonical or unchecked type" `llvm_unreachable` which trying to join `Array` and `[String]?` since `Array` is 'UnboundGenericType` which is "unchecked".